### PR TITLE
Websocket

### PR DESCRIPTION
**Pull request by Crypt-iQ**

### CODE REVIEW

**Comment by Giszmo** [comment]
@DanielKrawisz was about to figure out which license to pick for shuffle-java. This new dependency has a [dual-license](https://glassfish.java.net/public/CDDL+GPL_1_1.html) with GPL (potentially incompatible with what we pick. also ... v1? Seriously?) and  CDDL which I have no idea what it is.

File: shuffler/src/main/java/com/shuffle/p2p/WebsocketClientEndpoint.java [hunk 1]
```diff
@@ -0,0 +1,108 @@
+package com.shuffle.p2p;
```
Comment (DanielKrawisz) [inline shuffler/src/main/java/com/shuffle/p2p/WebsocketClientEndpoint.java]: IT looks like you don't really need this class at all, given that the only way you use it is just to call its newSession method. You could just cut out that one function and use that. If you really DO need this class, however, it should be a private inner class of WebsocketChannel. The user of this package will never use this class, so it shouldn't be presented to him.

File: shuffler/src/main/java/com/shuffle/p2p/WebsocketChannel.java [hunk 2]
```diff
@@ -2,20 +2,203 @@
...
+
+            //WebsocketPeer.WebsocketSession session = peer.newSession();
+
+            WebsocketClientEndpoint clientEndPoint = new WebsocketClientEndpoint(identity);
+            WebsocketPeer.WebsocketSession session = new WebsocketPeer(identity).new WebsocketSession(clientEndPoint.newSession());
+
+            return openSessions.put(identity, session);
+        }
+
+        public synchronized WebsocketPeer.WebsocketSession putOpenSessions(URI identity, javax.websocket.Session client) {
```
Comment (DanielKrawisz) [inline shuffler/src/main/java/com/shuffle/p2p/WebsocketChannel.java]: This function is not necessary. If you look at the corresponding function in TCPChannel, I wrote the comment "This is for creating a session that was initiated by a remote peer." That can't happen with the web socket channel.

File: shuffler/src/main/java/com/shuffle/p2p/WebsocketChannel.java [hunk 3]
```diff
@@ -2,20 +2,203 @@
...
+            return openSessions.get(identity);
+        }
+
+        public synchronized WebsocketPeer.WebsocketSession remove(URI identity) {
+            return openSessions.remove(identity);
+        }
+
+    }
+
+    OpenSessions openSessions = null;
```
Comment (DanielKrawisz) [inline shuffler/src/main/java/com/shuffle/p2p/WebsocketChannel.java]: It looks like you never initialize this variable. It should be initialized when the Channel.open() method is called. It should be reset to null when the corresponding Connection.close() method is called.

Comment (DanielKrawisz) [inline shuffler/src/main/java/com/shuffle/p2p/WebsocketChannel.java]: I just noticed that I made the same mistake in TCPChannel! Hahaha. You must have been really confused!

File: shuffler/src/main/java/com/shuffle/p2p/WebsocketChannel.java [hunk 4]
```diff
@@ -2,20 +2,203 @@
...
+            javax.websocket.Session session;
+
+            public WebsocketSession (javax.websocket.Session session) throws IOException {
+                this.session = session;
+            }
+
+            @Override
+            public boolean send(Bytestring message) {
+                try {
+                    session.getBasicRemote().sendText(message.toString()); // can't sendText a Bytestring, there's no toString for Bytestring?? Android Studio shows it, however.
```
Comment (DanielKrawisz) [inline shuffler/src/main/java/com/shuffle/p2p/WebsocketChannel.java]: message.bytes

File: shuffler/src/main/java/com/shuffle/p2p/WebsocketChannel.java [hunk 5]
```diff
@@ -2,20 +2,203 @@
...
+            @Override
+            public Peer<URI, Bytestring> peer() {
+                return WebsocketPeer.this;
+            }
+
+        }
+
+    }
+
+    private boolean running = false; // not necessary?
```
Comment (DanielKrawisz) [inline shuffler/src/main/java/com/shuffle/p2p/WebsocketChannel.java]: Running should be set to true when a connection object is created. Then it should be set to false when the corresponding connection object's close() function is called.

File: shuffler/src/main/java/com/shuffle/p2p/WebsocketChannel.java [hunk 6]
```diff
@@ -2,20 +2,203 @@
...
+            this.currentSession = session;
+        }
+
+        private WebsocketPeer setSession(javax.websocket.Session session) throws IOException {
+            currentSession = new WebsocketSession(session);
+            return this;
+        }
+
+        @Override
+        public synchronized Session<URI, Bytestring> openSession(Receiver<Bytestring> receiver) { // ignore receiver ?
```
Comment (DanielKrawisz) [inline shuffler/src/main/java/com/shuffle/p2p/WebsocketChannel.java]: No, you need to keep the receiver and pass new messages from the remote peer to it.

File: shuffler/src/main/java/com/shuffle/p2p/WebsocketChannel.java [hunk 7]
```diff
@@ -10,20 +10,259 @@
...
+import javax.websocket.ClientEndpoint;
+import javax.websocket.CloseReason;
+import javax.websocket.ContainerProvider;
+import javax.websocket.OnClose;
+import javax.websocket.OnOpen;
+import javax.websocket.Session;
+import javax.websocket.WebSocketContainer;
 
 /**
  * TODO
```
Comment (DanielKrawisz) [inline shuffler/src/main/java/com/shuffle/p2p/WebsocketChannel.java]: Don't need a TODO here anymore. Replace with appropriate comment explaining what this class is.

**Pull request merged**